libpromises/dbm_migration_lastseen.c

```diff
@@ -182,8 +188,96 @@ static bool LastseenMigrationVersion0(DBHandle *db)
     return !errors;
 }
 
+static bool LastseenMigrationVersion1(DBHandle *db)
+{
+    DBCursor *cursor;
+    if (!NewDBCursor(db, &cursor))
+    {
+        Log(LOG_LEVEL_ERR,
+            "Unable to create database cursor during lastseen migration");
+        return false;
+    }
+
+    char *key;
+    void *value;
+    int key_size, value_size;
+
+    // Iterate through all key-value pairs
+    while (NextDB(cursor, &key, &key_size, &value, &value_size))
+    {
+        if (key_size == 0)
+        {
+            Log(LOG_LEVEL_WARNING,
+                "Found zero-length key during lastseen migration");
+            continue;
+        }
+
+        // Only look for old KeyHostSeen entries
+        if (key[0] != 'q')
+        {
+            // Warn about completely unexpected keys
+            if ((key[0] != 'k') && (key[0] != 'a'))
+            {
+                Log(LOG_LEVEL_WARNING,
+                    "Found unexpected key '%s' during lastseen migration",
+                    key);
+            }
+            continue;
+        }
+
+        KeyHostSeen1 *old_value = value;
+        KeyHostSeen new_value = {
+            .acknowledged = true, // We don't know, assume yes
+            .lastseen = old_value->lastseen,
+            .Q = old_value->Q,
+        };
+
+        if (!DBCursorDeleteEntry(cursor))
+        {
+            Log(LOG_LEVEL_ERR,
+                "Unable to delete version 1 of entry for key '%s' during lastseen migration",
+                key);
+            if (DeleteDBCursor(cursor) == false)
+            {
+                Log(LOG_LEVEL_ERR,
+                    "Unable to close cursor during lastseen migration");
+            }
+            return false;
+        }
+
+        if (!WriteDB(db, key, &new_value, sizeof(new_value)))
```

Comment:
To quote [our "docs"](https://github.com/cfengine/core/blob/master/libpromises/dbm_api.h#L99):

```
 * Don't use WriteDB/DeleteDB while iterating database, it will result in
 * deadlock. Use cursor-specific operations instead. They work on the current
 * key.
 */
```